Feat/grpc tls

fabriziomoscon opened the pull request:
Adding TLS for grpc executor.

Remove unused option plaintext.
Add move options to grpc executor for TLS connections:
- tls_root_ca
- tls_client_cert
- tls_client_key
- ignore_verify_ssl

Is there any existing test for grpc? I didn't find it.

fsamin commented:
Hello,
Thanks for your pull request 👍 
Sadly, we don't have tests for the GRPC executor, could we ask you to add a few tests ?

ovh-cds commented:
CDS Report build-venom-a#706.0 ✘  
* Build
  * Build ✔ 
* Tests
  * Acceptance Tests ✘

fabriziomoscon commented:
@fsamin yes I will add tests. Thanks.

fabriziomoscon commented:
@fsamin I had a look at the other tests, and it is not clear to me how I would add this test, could you please suggest your preferred way.
Usually, we test this with a go unit test, setup creates local files for pki, the test spawn a TLS server and connects a TLS client with mutual TLS.

fabriziomoscon commented:
@fsamin @yesnault I have added a docker grpc server with TLS as `venom-grpc.cid`, which works. But I don't understand how to run venom against it. I have added a target `grpc-test` which uses `venom_wrapper.sh` to run my test locally but I receive an error:
```
make: *** [grpc-test] Error 127
```

Please let me know how would I run this test

fsamin commented:
Our integration tests follows the following process:

To starts all tests dependencies. We are using `make start-test-stack` to start all dependencies as docker containers. We are using .cid files to store containers ID. Your `venom-grpc.cid` target seems fine.

Once all the dependencies are up and running, we compile a specific binary of venom (with the golang tests intrumentation to be able to gather code coverage). We are using `make build-test-binary-docker`. It produces a `venom.test` file.

To run the tests we are using `make run-test`. Under the hood, it relies on venom_wrappers.sh. It runs all yaml files in the tests directory and if needed you can add `--var-from-file` flag to specify specific variables to run your tests against your grpc server.

kravcs commented:
Thanks @fsamin for the instructions, the integration tests for grpc TLS are working fine from our perspective.
We also added the ability to setup TLS options with PEM files in similar way as you did for http executor.
Please let us know if everything is fine from your side as well.

fabriziomoscon commented on tests/Makefile:
```diff
@@ -57,8 +75,9 @@ build-test-binary:
-	VENOM_VAR_MY_ENVAR=foo ./venom_wrapper.sh run -vv --format=xml --output-dir=. --lib-dir=./lib_custom --var='array_from_var=["biz","buz"]' --var-from-file ./kafka/testVariables.yml  --var-from-file ./vars/vars.yml ./*.yml ./assertions/*.yml && \
+run-test: generate-venom-pki
+	VENOM_VAR_MY_ENVAR=foo  ./venom_wrapper.sh run \
+	-vv --format=xml --output-dir=. --lib-dir=./lib_custom --var='array_from_var=["biz","buz"]' --var-from-file ./kafka/testVariables.yml --var-from-file=$(PKI_VAR_FILE) --var-from-file ./vars/vars.yml ./*.yml ./assertions/*.yml && \
```
this `--var-from-file=$(PKI_VAR_FILE)` is needed to generate the yaml formatted PEM import file used in the test.

fabriziomoscon commented:
@fsamin just wanted to confirm that, as far as we can tell, we completed all the fixes for this PR. We are looking forward to read your new instructions.
Thank you

ovh-cds commented:
CDS Report build-venom-a#792.0 ✘  
* Build
  * Build ✔ 
* Tests
  * Acceptance Tests ✘

fsamin commented on tests/pki.mk:
```diff
@@ -0,0 +1,72 @@
+	openssl genrsa \
+		-out $(PKI_DIR)/server.key 2048
+
+	# create a server Certificate Signing Request
```
This step fails with the following error:

```
mkdir -p ./grpc/pki
# create a certificate authority (CA) that both the client and server trust.
# The CA is just a public and private key with the public key wrapped up in a self-signed X.509 certificate.
openssl req \
	-new \
	-x509 \
	-nodes \
	-days 365 \
	-subj '/C=GB/O=Example/OU=TeamA/CN=ca.example.com' \
	-keyout ./grpc/pki/ca.key \
	-out ./grpc/pki/ca.crt
3m54s234ms
Generating a RSA private key
..........................................+++++
...+++++
writing new private key to './grpc/pki/ca.key'
-----
unable to find 'distinguished_name' in config
problems making Certificate Request
139874421728576:error:0E06D06C:configuration file routines:NCONF_get_string:no value:../crypto/conf/conf_lib.c:273:group=req name=distinguished_name
make: *** [pki.mk:5: grpc/pki] Error 1
```

Could you handle this ?

fabriziomoscon replied:
Is the CI on linux?

yesnault replied:
yes, ubuntu-20.04

fabriziomoscon replied:
I quickly searched and found this: https://superuser.com/questions/947061/openssl-unable-to-find-distinguished-name-in-config
It looks like `openssl req` is not able to find a config value...
could you please let me know which openSSL version you run on ubuntu?

fabriziomoscon replied:
Also, we don't use `-subj` in test, so perhaps that could be removed and see if the `distinguished_name` issue goes away

sarunask replied:
I have tried to:
1. Run in Docker ubuntu:20.04
2. Install openssl with `apt update && apt install --yes openssl`
3. Run 2 commands and they were successful:
```
root@7bd2261de939:~# mkdir -p ./grpc/pki
root@7bd2261de939:~# openssl req \
> -new \
> -x509 \
> -nodes \
> -days 365 \
> -subj '/C=GB/O=Example/OU=TeamA/CN=ca.example.com' \
> -keyout ./grpc/pki/ca.key \
> -out ./grpc/pki/ca.crt
Generating a RSA private key
.........................+++++
...................................................................................................................+++++
writing new private key to './grpc/pki/ca.key'
-----
```
Question would be: Are you using vanilla Ubuntu 20.04 Docker image? Does it have openssl installed ? Or you use some other custom Docker image with openssl installed in some other way ?

sarunask replied:
It seems your openssl can't find `/usr/lib/ssl/openssl.cnf` thus question is - is this file available on your Docker image?

yesnault replied:
@sarunask @fabriziomoscon  openssl fixed on your worker model. Tests are now ok. Can you rebase your PR please?

fabriziomoscon replied:
done

ovh-cds commented:
CDS Report build-venom-a#792.1 ✘  
* Build
  * Build ✔ 
* Tests
  * Acceptance Tests ✘

ovh-cds commented:
CDS Report build-venom-a#795.0 ✘  
* Build
  * Build ✔ 
* Tests
  * Acceptance Tests ✘

ovh-cds commented:
CDS Report build-venom-a#792.2 ■   
* Build
  * Build ■

ovh-cds commented:
CDS Report build-venom-a#792.2 ■   
* Build
  * Build ■

fabriziomoscon commented:
Thank you
